homeassistant/components/integration/sensor.py

```diff
@@ -150,17 +150,29 @@ def calc_integration(event):
             old_state = event.data.get("old_state")
             new_state = event.data.get("new_state")
 
+            # We may want to update our state before an early return,
+            # based on the source sensor's unit_of_measurement
+            # or device_class.
+            update_state = False
+
             if self._unit_of_measurement is None:
                 unit = new_state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
-                self._unit_of_measurement = self._unit_template.format(
-                    "" if unit is None else unit
-                )
+                if unit is not None:
+                    self._unit_of_measurement = self._unit_template.format(
+                        "" if unit is None else unit
```

Comment:
This does not make sense... you are inside a block where `unit is not None`

Reply:
Good point.  I was just wrapped the existing assignment into a block to protect it from unit == None.  Now that it's protected, we can assume it's not None during the assignment.  
I'll remove the redundant check.

Reply:
The problem with the original line, just to be clear, is that when unit was None, there was still an assignment to self._unit_of_measurement, placing a "" if unit was None.  Therefore self._unit_of_measurement was not None anymore and this assignment didn't execute on the next pass (when unit IS populated).

Reply:
This has been updated.